File: NOIP2008复赛/CZOI2008铜组/day6/solution/knight/knight_a.pas
const
  MAXN = 10;
  MAXM = 10;
  MAX = MAXN * MAXM;
  delt: array[1..8,1..2] of integer = ((1, -2), (2, -1), (2, 1), (1, 2), (-1, 2), (-2, 1), (-2, -1), (-1, -2));

var
  n, m, sx, sy: longint;
  x, y, top, i: longint;
  stack: array[1..MAX] of longint;
  board: array[1..MAXN, 1..MAXM] of boolean;

begin
  assign(input, 'knight.in');
  reset(input);
  read(n, m, sx, sy);
  close(input);
  fillchar(board, sizeof(board), 0);
  x := sx; y := sy;
  board[x, y] := true;
  top := 1;
  stack[top] := 0;
  while (top >= 1) and (top < n * m) do
  begin
    inc(stack[top]);
    if (stack[top] <= 8) then
    begin
      inc(x, delt[stack[top], 1]);
      inc(y, delt[stack[top], 2]);
      if (x <= 0) or (x > n) or (y <= 0) or (y > m) or (board[x, y]) then
      begin
        dec(x, delt[stack[top], 1]);
        dec(y, delt[stack[top], 2])
      end
      else
      begin
        board[x, y] := true;
        inc(top);
        stack[top] := 0
      end
    end
    else
    begin
      board[x, y] := false;
      dec(top);
      dec(x, delt[stack[top], 1]);
      dec(y, delt[stack[top], 2])
    end
  end;
  assign(output, 'knight.out');
  rewrite(output);
  if (top > 1) then
  begin
    write('(', sx, ',', sy, ')');
    for i := 1 to n * m - 1 do
    begin
      inc(sx, delt[stack[i], 1]);
      inc(sy, delt[stack[i], 2]);
      write('->(', sx, ',', sy, ')')
    end;
    writeln
  end
  else
    writeln('No Answer!');
  close(output)
end.

File: NOIP2008复赛/CZOI2008铜组/day6/solution/knight/knight_a.c
#include <stdio.h>
#include <stdlib.h>

#define MAXN 10
#define MAXM 10
#define MAX MAXN*MAXM

const
int delt[8][2] = {{1, -2}, {2, -1}, {2, 1}, {1, 2}, {-1, 2}, {-2, 1}, {-2, -1}, {-1, -2}};

int main() {
  int n, m, sx, sy;
  int x, y, top, i;
  int stack[MAX], board[MAXN][MAXM];
  FILE *fin, *fout;
  fin = fopen("knight.in", "r");
  fscanf(fin, "%d %d %d %d", &n, &m, &sx, &sy);
  fclose(fin);
  memset(board, 0, sizeof(board));
  x = sx - 1; y = sy - 1;
  board[x][y] = 1;
  top = 0;
  stack[top] = -1;
  while ((top >= 0) && (top < n * m - 1)) {
    stack[top]++;
    if (stack[top] < 8) {
      x = x + delt[stack[top]][0];
      y = y + delt[stack[top]][1];
      if ((x < 0) || (x >= n) || (y < 0) || (y >= m) || (board[x][y] == 1)) {
        x = x - delt[stack[top]][0];
        y = y - delt[stack[top]][1];
      }
      else {
        board[x][y] = 1;
        stack[++top] = -1;
      }
    }
    else {
      board[x][y] = 0;
      top--;
      x = x - delt[stack[top]][0];
      y = y - delt[stack[top]][1];
    }
  }
  fout = fopen("knight.out", "w");
  if (top > 0) {
    fprintf(fout, "(%d,%d)", sx, sy);
    for (i = 0; i < n * m - 1; i++) {
      sx = sx + delt[stack[i]][0];
      sy = sy + delt[stack[i]][1];
      fprintf(fout, "->(%d,%d)", sx, sy);
    }
    fprintf(fout, "\n");
  }
  else
    fprintf(fout, "No Answer!\n");
  fclose(fout);
  return 0;
}
